cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/class-sync.php

```diff
@@ -119,66 +163,654 @@ public function is_synced( $post_id ) {
 	/**
 	 * Generate a signature based on whats required for a full sync.
 	 *
-	 * @param int $post_id The post id to generate a signature for.
+	 * @param int  $attachment_id The Attachment id to generate a signature for.
+	 * @param bool $cache         Flag to specify if a cached signature is to be used or build a new one.
 	 *
 	 * @return string|bool
 	 */
-	public function generate_signature( $post_id ) {
-		$upload = $this->managers['push']->prepare_upload( $post_id );
-		// Check if has an error (ususally due to file quotas).
-		if ( is_wp_error( $upload ) ) {
-			$this->plugin->components['media']->get_post_meta( $post_id, self::META_KEYS['sync_error'], $upload->get_error_message() );
+	public function generate_signature( $attachment_id, $cache = true ) {
+		static $signatures = array(); // cache signatures.
+		if ( ! empty( $signatures[ $attachment_id ] ) && true === $cache ) {
+			$return = $signatures[ $attachment_id ];
+		} else {
+			$return = $this->sync_base( $attachment_id );
+			if ( ! is_wp_error( $return ) ) {
+				$return                       = array_map(
```

Comment:
The spaces here might not be required

Reply:
IT's got spaces because it's directly inline with the line below it.
I'll add a comment explaining it's adding to the cache, which will remove the spaces.